### PR TITLE
Cuda matrix conversion csr ell

### PR DESCRIPTION
**Pull request by fritzgoebel**
Cuda and reference conversion routines between CSR and ELL formats.

__TODO:__
- [x] Rebase

### CODE REVIEW

File: cuda/matrix/ell_kernels.cu [hunk 1]
```diff
@@ -405,6 +407,190 @@ GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
...
+    IndexType *__restrict__ result_row_ptrs,
+    IndexType *__restrict__ result_col_idxs,
+    ValueType *__restrict__ result_values)
+{
+    const auto tidx = threadIdx.x + blockDim.x * blockIdx.x;
+
+    if (tidx < num_rows) {
+        auto write_to = result_row_ptrs[tidx];
+        for (auto i = 0; i < max_nnz_per_row; i++) {
+            if (source_values[tidx + stride * i] != zero<ValueType>()) {
```
Comment (thoasm) [inline cuda/matrix/ell_kernels.cu]: Here, you could also save the index calculation inside a separate variable (the result of `tidx + stride * i`).

File: cuda/matrix/ell_kernels.cu [hunk 2]
```diff
@@ -405,6 +407,190 @@ GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
...
+    auto num_rows = result->get_size()[0];
+
+    auto row_ptrs = result->get_row_ptrs();
+    auto col_idxs = result->get_col_idxs();
+    auto values = result->get_values();
+
+    const auto stride = source->get_stride();
+    const auto max_nnz_per_row = source->get_num_stored_elements_per_row();
+
+    const auto rows_per_block =
```
Comment (thoasm) [inline cuda/matrix/ell_kernels.cu]: Nit: Would make this `constexpr`
```suggestion
    constexpr auto rows_per_block =
```

File: cuda/test/matrix/csr_kernels.cpp [hunk 3]
```diff
@@ -86,8 +91,8 @@ class Csr : public ::testing::Test {
     void set_up_apply_data(std::shared_ptr<Mtx::strategy_type> strategy)
     {
         mtx = Mtx::create(ref, strategy);
-        mtx->copy_from(gen_mtx<Vec>(532, 231, 1));
-        expected = gen_mtx<Vec>(532, 1, 1);
+        mtx->copy_from(gen_mtx<Vec>(512, 231, 1));
+        expected = gen_mtx<Vec>(512, 1, 1);
```
Comment (thoasm) [inline cuda/test/matrix/csr_kernels.cpp]: Why the change from 532 to 512?
532 seems like a good number to ensure that more than one thread block is used.

File: cuda/test/matrix/csr_kernels.cpp [hunk 4]
```diff
@@ -41,10 +41,15 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #include <gtest/gtest.h>
 
 
+#include "core/matrix/csr_kernels.hpp"
+#include "core/matrix/dense_kernels.hpp"
+
+
```
Comment (thoasm) [inline cuda/test/matrix/csr_kernels.cpp]: Private Ginkgo header (`core/test/utils.hpp` should be one of them, and also written with `""` instead of `<>`) should be below the public Ginkgo header.

File: reference/matrix/csr_kernels.cpp [hunk 5]
```diff
@@ -165,6 +166,46 @@ GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
...
+}
+
+GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
+    GKO_DECLARE_CSR_CONVERT_TO_ELL_KERNEL);
+
+
+template <typename ValueType, typename IndexType>
+void move_to_ell(std::shared_ptr<const ReferenceExecutor> exec,
+                 matrix::Ell<ValueType, IndexType> *result,
+                 matrix::Csr<ValueType, IndexType> *source) GKO_NOT_IMPLEMENTED;
```
Comment (thoasm) [inline reference/matrix/csr_kernels.cpp]: We should definitively not have any `NOT_IMPLEMENTED` kernels inside reference, so I would strongly suggest to remove this kernel all together.

File: reference/test/matrix/csr_kernels.cpp [hunk 6]
```diff
@@ -230,6 +231,20 @@ TEST_F(Csr, MovesToCoo)
...
+    auto ell_mtx = gko::matrix::Ell<>::create(mtx->get_executor());
+    auto dense_mtx = gko::matrix::Dense<>::create(mtx->get_executor());
+    auto ref_dense_mtx = gko::matrix::Dense<>::create(mtx->get_executor());
+
+    mtx->convert_to(ell_mtx.get());
+    ell_mtx->convert_to(dense_mtx.get());
+
+    mtx->convert_to(ref_dense_mtx.get());
+
+    GKO_ASSERT_MTX_NEAR(dense_mtx.get(), ref_dense_mtx.get(), 0.0);
```
Comment (thoasm) [inline reference/test/matrix/csr_kernels.cpp]: I don't think this test is enough, you also need one test to check if the values are set appropriately by hand (when just converting from CSR to ELL).

File: reference/test/matrix/ell_kernels.cpp [hunk 7]
```diff
@@ -323,4 +324,28 @@ TEST_F(Ell, MovesWithStrideToDense)
...
+
+TEST_F(Ell, ConvertsWithStrideToCsr)
+{
+    auto csr_mtx = gko::matrix::Csr<>::create(mtx2->get_executor());
+    auto dense_ref = gko::matrix::Dense<>::create(mtx2->get_executor());
+
+    mtx2->convert_to(csr_mtx.get());
+    csr_mtx->convert_to(dense_ref.get());
+
+    GKO_ASSERT_MTX_NEAR(mtx1, dense_ref, 0.0);
```
Comment (thoasm) [inline reference/test/matrix/ell_kernels.cpp]: Same here to both added tests: You also need to check the result manually if it works the way you want.

File: cuda/matrix/csr_kernels.cu [hunk 8]
```diff
@@ -824,6 +917,95 @@ GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
...
+    const auto num_threads = blockDim.x * gridDim.x;
+
+    __shared__ IndexType block_result[default_block_size];
+
+    block_result[threadIdx.x] =
+        (tidx < num_rows) ? row_ptrs[tidx + 1] - row_ptrs[tidx] : 0;
+
+    for (auto i = num_threads; i < num_rows; i += num_threads) {
+        block_result[threadIdx.x] =
+            max(block_result[threadIdx.x], row_ptrs[i + 1] - row_ptrs[i]);
```
Comment (yhmtsai) [inline cuda/matrix/csr_kernels.cu]: It seems that you take block_result to compare the same things in different threads.
I guess you mean `for (auto i = num_threads + tidx; ....)` here

File: cuda/components/prefix_sum.cuh [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+template <int block_size, typename ValueType>
+__global__ __launch_bounds__(block_size) void start_prefix_sum(
+    size_type num_elements, ValueType *__restrict__ elements,
+    ValueType *__restrict__ block_sum)
+{
+    const auto tidx = threadIdx.x + blockDim.x * blockIdx.x;
+    __shared__ size_type prefix_helper[block_size];
+    prefix_helper[threadIdx.x] =
+        (tidx < num_elements) ? elements[tidx] : zero<ValueType>();
+    __syncthreads();
```
Comment (thoasm) [inline cuda/components/prefix_sum.cuh]: Since PR #230, we try to prevent the usage of `__syncthreads()` and similar by using [cooperative groups for everything](https://github.com/ginkgo-project/ginkgo/blob/develop/cuda/components/cooperative_groups.cuh).
To do that, you need to `#include "cuda/components/cooperative_groups.cuh"` and do:
```c++
group::this_thread_block().sync();
```

Here, because you use it multiple times, it is better to store the block in a variable and sync with it:
```
auto this_block = group::this_thread_block();
// ...
this_block.sync();
```
And replace `__syncthreads();` with `this_block.sync();`

File: cuda/components/reduction.cuh [hunk 10]
```diff
@@ -146,6 +146,33 @@ __device__ void reduce(const Group &__restrict__ group,
...
+                             Operator reduce_op = Operator{})
+{
+    const auto tidx = threadIdx.x + blockIdx.x * blockDim.x;
+    auto thread_result = zero<ValueType>();
+    for (auto i = tidx; i < size; i += blockDim.x * gridDim.x) {
+        thread_result = reduce_op(thread_result, source[i]);
+    }
+    result[threadIdx.x] = thread_result;
+
+    __syncthreads();
```
Comment (thoasm) [inline cuda/components/reduction.cuh]:
```suggestion
    group::this_thread_block().sync();
```
You can actually store the current block and use it in the reduction below again:
```c++
auto this_block = group::this_thread_block();
this_block.sync();
reduce(this_block, result, reduce_op);
```

File: core/matrix/ell_kernels.hpp [hunk 11]
```diff
@@ -61,13 +62,43 @@ namespace kernels {
...
-    GKO_DECLARE_ELL_CONVERT_TO_DENSE_KERNEL(ValueType, IndexType)
+#define GKO_DECLARE_ELL_CONVERT_TO_CSR_KERNEL(ValueType, IndexType)  \
+    void convert_to_csr(std::shared_ptr<const DefaultExecutor> exec, \
+                        matrix::Csr<ValueType, IndexType> *result,   \
+                        const matrix::Ell<ValueType, IndexType> *source)
+
+#define GKO_DECLARE_ELL_MOVE_TO_CSR_KERNEL(ValueType, IndexType)  \
+    void move_to_csr(std::shared_ptr<const DefaultExecutor> exec, \
+                     matrix::Csr<ValueType, IndexType> *result,   \
+                     const matrix::Ell<ValueType, IndexType> *source)
```
Comment (thoasm) [inline core/matrix/ell_kernels.hpp]: I don't think you have `move_to_csr` anymore, so you can delete it here as well.

File: cuda/matrix/csr_kernels.cu (hunk 8 above)
Comment (thoasm) [reply cuda/matrix/csr_kernels.cu]: I think this part of the code is now gone (replaced by a `reduce` call if I remember correctly), but I am not 100% sure.
@fritzgoebel do you remember where this part was actually from?

Comment (fritzgoebel) [reply cuda/matrix/csr_kernels.cu]: Yes, this part of the code is gone.

File: core/matrix/ell.cpp [hunk 12]
```diff
@@ -120,6 +125,29 @@ void Ell<ValueType, IndexType>::move_to(Dense<ValueType> *result)
...
+    size_type num_stored_elements = 0;
+    exec->run(ell::make_count_nonzeros(this, &num_stored_elements));
+
+    auto tmp = Csr<ValueType, IndexType>::create(exec, this->get_size(),
+                                                 num_stored_elements);
+    exec->run(ell::make_convert_to_csr(tmp.get(), this));
+
+    tmp->move_to(result);
+}
+
```
Comment (yhmtsai) [inline core/matrix/ell.cpp]: need one more empty line

File: cuda/matrix/dense_kernels.cu [hunk 13]
```diff
@@ -827,35 +760,9 @@ void move_to_sellp(std::shared_ptr<const CudaExecutor> exec,
 GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
     GKO_DECLARE_DENSE_MOVE_TO_SELLP_KERNEL);
 
-
```
Comment (yhmtsai) [inline cuda/matrix/dense_kernels.cu]: should keep this empty line

File: cuda/test/matrix/csr_kernels.cpp [hunk 14]
```diff
@@ -343,5 +348,34 @@ TEST_F(Csr, ConvertToDenseIsEquivalentToRef)
...
+    auto ell_mtx = gko::matrix::Ell<>::create(ref);
+    auto dell_mtx = gko::matrix::Ell<>::create(cuda);
+
+    mtx->convert_to(ell_mtx.get());
+    dmtx->convert_to(dell_mtx.get());
+
+    GKO_ASSERT_MTX_NEAR(ell_mtx.get(), dell_mtx.get(), 1e-14);
+}
+
```
Comment (yhmtsai) [inline cuda/test/matrix/csr_kernels.cpp]: need one more line

File: reference/test/matrix/csr_kernels.cpp [hunk 15]
```diff
@@ -229,6 +252,16 @@ TEST_F(Csr, MovesToCoo)
     assert_equal_to_mtx(coo_mtx.get());
 }
```
Comment (yhmtsai) [inline reference/test/matrix/csr_kernels.cpp]: need one more line

File: cuda/test/matrix/csr_kernels.cpp [hunk 16]
```diff
@@ -343,5 +348,34 @@ TEST_F(Csr, ConvertToDenseIsEquivalentToRef)
     GKO_ASSERT_MTX_NEAR(dense_mtx.get(), ddense_mtx.get(), 1e-14);
 }
```
Comment (yhmtsai) [inline cuda/test/matrix/csr_kernels.cpp]: need one more line

File: omp/matrix/csr_kernels.cpp [hunk 17]
```diff
@@ -232,6 +242,15 @@ void conj_transpose(std::shared_ptr<const OmpExecutor> exec,
...
 
+
+template <typename ValueType, typename IndexType>
+void calculate_max_nnz_per_row(std::shared_ptr<const OmpExecutor> exec,
+                               const matrix::Csr<ValueType, IndexType> *source,
+                               size_type *result) GKO_NOT_IMPLEMENTED;
+
+GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
+    GKO_DECLARE_CSR_CALCULATE_MAX_NNZ_PER_ROW_KERNEL);
+
```
Comment (yhmtsai) [inline omp/matrix/csr_kernels.cpp]: need one more empty line

File: cuda/components/prefix_sum.cuh [hunk 18]
```diff
@@ -0,0 +1,144 @@
...
+
+/**
+ * @internal
+ * Calculates the prefix sum of `elements` inside `block_size` sized blocks
+ * in-place. Can be used together with `finalize_prefix_sum` to calculate the
+ * prefix-sum of an array. `block_size` must be the thread block size used when
+ * calling this kernel, which also dictates the size of blocks where the
+ * in-place prefix sum will be calculated `block_sum` is used to store the total
+ * sum of each block. It requires at least `ceildiv(num_elements, block_size)`
+ * elements
```
Comment (tcojean) [inline cuda/components/prefix_sum.cuh]: As this looks like duplicated from the other PR, whichever fixes you did need to also be done here.

File: cuda/components/prefix_sum.cuh [hunk 19]
```diff
@@ -0,0 +1,144 @@
...
+/**
+ * @internal
+ * Increases the value of each entry of `elements` by the total sum of all
+ * preceding blocks of size `block_size`. Can be used together with
+ * `start_prefix_sum` to calculate the prefix-sum of an array. `block_size` must
+ * be the thread block size used when calling this kernel, which also dictates
+ * the size of blocks where the in-place prefix sum will be calculated by
+ * `start_prefix_sum`. `block_sum` is used by `start_prefix_sum` to store the
+ * total sum of each block. It requires at least `ceildiv(num_elements,
+ * block_size)` elements
```
Comment (tcojean) [inline cuda/components/prefix_sum.cuh]: Same here

File: cuda/components/prefix_sum.cuh [hunk 20]
```diff
@@ -0,0 +1,144 @@
...
+ * elements
+ */
+template <int block_size, typename ValueType>
+__global__ __launch_bounds__(block_size) void start_prefix_sum(
+    size_type num_elements, ValueType *__restrict__ elements,
+    ValueType *__restrict__ block_sum)
+{
+    const auto tidx = threadIdx.x + blockDim.x * blockIdx.x;
+    __shared__ size_type prefix_helper[block_size];
+    prefix_helper[threadIdx.x] =
```
Comment (tcojean) [inline cuda/components/prefix_sum.cuh]: same here

File: cuda/components/prefix_sum.cuh (hunk 18 above)
Comment (thoasm) [reply cuda/components/prefix_sum.cuh]: I would say he changes it in one PR (whichever it is), and when we merge it, he can rebase it, so it is also present in the other one.

**Comment by tcojean** [comment]
After merging #225 you have a conflict here so you definitely have to rebase/update.

**Comment by tcojean** [comment]
This PR conflicted a lot with #279 sadly, so you definitely need to rebase, once more...
